dex/encrypt/snacl.go

```diff
@@ -0,0 +1,269 @@
+	var nonce [NonceSize]byte
+	_, err := io.ReadFull(prng, nonce[:])
+	if err != nil {
+		return nil, fmt.Errorf("CryptoKey.Encrypt: %v", err)
+	}
+	blob := secretbox.Seal(nil, in, &nonce, (*[KeySize]byte)(ck))
+	return append(nonce[:], blob...), nil
+}
+
+// Decrypt decrypts the passed data.  The must be the output of the Encrypt
+// function.
+func (ck *CryptoKey) Decrypt(in []byte) ([]byte, error) {
+	if len(in) < NonceSize {
+		return nil, fmt.Errorf("missing nonce")
+	}
+
+	var nonce [NonceSize]byte
+	copy(nonce[:], in[:NonceSize])
+	blob := in[NonceSize:]
+
+	opened, ok := secretbox.Open(nil, blob, &nonce, (*[KeySize]byte)(ck))
+	if !ok {
+		return nil, fmt.Errorf("faile dto open")
+	}
+
+	return opened, nil
+}
+
+// Zero clears the key by manually zeroing all memory.  This is for security
+// conscience application which wish to zero the memory after they've used it
+// rather than waiting until it's reclaimed by the garbage collector.  The
+// key is no longer usable after this call.
+func (ck *CryptoKey) Zero() {
+	*ck = [KeySize]byte{}
+}
+
+// GenerateCryptoKey generates a new crypotgraphically random key.
+func GenerateCryptoKey() (*CryptoKey, error) {
+	var key CryptoKey
+	_, err := io.ReadFull(prng, key[:])
+	if err != nil {
+		return nil, err
+	}
+
+	return &key, nil
+}
+
+// Parameters are not secret and can be stored in plain text.
+type Parameters struct {
+	Salt   [KeySize]byte
+	Digest [sha256.Size]byte
+	N      int
+	R      int
+	P      int
+}
+
+// SecretKey houses a crypto key and the parameters needed to derive it from a
+// passphrase.  It should only be used in memory.
+type SecretKey struct {
+	Key        *CryptoKey
+	Parameters Parameters
+}
+
+func zero(b []byte) {
+	for i := range b {
+		b[i] = 0
+	}
+}
+
+// deriveKey fills out the Key field.
+func (sk *SecretKey) deriveKey(password *[]byte) error {
+	key, err := scrypt.Key(*password, sk.Parameters.Salt[:],
+		sk.Parameters.N,
+		sk.Parameters.R,
+		sk.Parameters.P,
+		len(sk.Key))
+	if err != nil {
+		return err
+	}
+	copy(sk.Key[:], key)
+	zero(key)
+
+	// I'm not a fan of forced garbage collections, but scrypt allocates a
+	// ton of memory and calling it back to back without a GC cycle in
+	// between means you end up needing twice the amount of memory.  For
+	// example, if your scrypt parameters are such that you require 1GB and
+	// you call it twice in a row, without this you end up allocating 2GB
+	// since the first GB probably hasn't been released yet.
+	debug.FreeOSMemory()
+
+	// I'm not a fan of forced garbage collections, but scrypt allocates a
+	// ton of memory and calling it back to back without a GC cycle in
+	// between means you end up needing twice the amount of memory.  For
+	// example, if your scrypt parameters are such that you require 1GB and
+	// you call it twice in a row, without this you end up allocating 2GB
+	// since the first GB probably hasn't been released yet.
+	debug.FreeOSMemory()
```

Comment:
Copied directly from dcrwallet, but double `FreeOSMemory` seems extraneous?

Reply:
Yeah, this must be unintentional.